### PR TITLE
Greg leaves the umbrella

### PR DESCRIPTION
**Pull request by gcolvin**
Refactoring to support configuration of optimizations, and code for same.

### CODE REVIEW

File: libevm/VM.cpp [hunk 1]
```diff
@@ -74,14 +50,14 @@ template <class S> S modWorkaround(S const& _a, S const& _b)
 void VM::onOperation()
 {
 	if (*m_onOp)
-		(*m_onOp)(++m_nSteps, m_PC, m_inst,
+		(*m_onOp)(++m_nSteps, PC, INST,
```
Comment (winsvega) [inline libevm/VM.cpp]: why without m_ ?

File: libevm/VM.cpp [hunk 2]
```diff
@@ -134,15 +122,12 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 {
 	m_io_gas = &io_gas;
 	m_ext = &_ext;
-	m_schedule = &m_ext->evmSchedule();
+   m_schedule = &m_ext->evmSchedule();
```
Comment (winsvega) [inline libevm/VM.cpp]: tabulation

File: libevm/VM.cpp [hunk 3]
```diff
@@ -117,14 +93,26 @@ void VM::updateMem()
...
-	m_newMemSize = memNeed(*m_SP, *(m_SP - 1));
+	unsigned n = (unsigned)INST - (unsigned)Instruction::LOG0;
+	m_runGas = toUint64(m_schedule->logGas + m_schedule->logTopicGas * n + u512(m_schedule->logDataGas) * *(SP - 1));
+	m_newMemSize = memNeed(*SP, *(SP - 1));
 	updateMem();
 }
 
+void VM::fetchInstruction()
+{
+		INST =  Instruction(PC < m_code_vector.size() ? m_code[PC] : 0);
```
Comment (chriseth) [inline libevm/VM.cpp]: Indentation is wrong and double space after `=`

File: libevm/VMOpt.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	{
+		byte inst = m_code[i];
+
+		if ((byte)Instruction::PUSH1 <= inst && inst <= (byte)Instruction::PUSH32)
+		{
+			byte n = inst - (byte)Instruction::PUSH1 + 1;
+
+			// decode pushed bytes to integral value
+			u256 val = m_code[i+1];
+			for (uint64_t j = i+2, m = n; --m; ++j)
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: Lots of potential for off-by-one's, also there should already be code somewhere that does this. Can we re-use it?

File: libevm/VMOpt.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+			}			
+#endif
+
+#ifdef EVM_REPLACE_CONST_JUMP
+			// replace JUMP or JUMPI to constant location with JUMPV or JUMPVI
+			size_t ii = i + n + 1;
+			byte op = m_code[ii];
+			if (op == (byte)Instruction::JUMP)
+			{
+				if (verifyJumpDest(val))
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: This has logarithmic complexity.

File: libevmcore/Instruction.cpp [hunk 6]
```diff
@@ -82,6 +82,8 @@ const std::map<std::string, Instruction> dev::eth::c_instructions =
 	{ "SSTORE", Instruction::SSTORE },
 	{ "JUMP", Instruction::JUMP },
 	{ "JUMPI", Instruction::JUMPI },
+	{ "JUMPV", Instruction::JUMP },
```
Comment (chriseth) [inline libevmcore/Instruction.cpp]: I think it would be better to add these to a separate table - they are not official opcodes.

File: libevmcore/Instruction.h [hunk 7]
```diff
@@ -127,16 +127,16 @@ enum class Instruction: uint8_t
...
 	DUP1 = 0x80,		///< copies the highest item in the stack to the top of the stack
-	DUP2,				///< copies the second highest item in the stack to the top of the stack
-	DUP3,				///< copies the third highest item in the stack to the top of the stack
-	DUP4,				///< copies the 4th highest item in the stack to the top of the stack
-	DUP5,				///< copies the 5th highest item in the stack to the top of the stack
-	DUP6,				///< copies the 6th highest item in the stack to the top of the stack
-	DUP7,				///< copies the 7th highest item in the stack to the top of the stack
-	DUP8,				///< copies the 8th highest item in the stack to the top of the stack
-	DUP9,				///< copies the 9th highest item in the stack to the top of the stack
+	DUP2,					///< copies the second highest item in the stack to the top of the stack
```
Comment (chriseth) [inline libevmcore/Instruction.h]: Indentation.

File: libevm/VMOpt.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+				val = (val << 8) | m_code[j];
+
+#ifdef EVM_USE_CONSTANT_POOL			
+			// add value to constant pool and replace PUSHn with PUSHC if room
+			if (1 < n)
+			{
+				int pool_off = pool_constant(val);
+				if (0 <= pool_off && pool_off < 256)
+				{
+					m_code[i] = (byte)Instruction::PUSHC;
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: If you change the code here to introduce new opcodes, you also have to scan the code to check that such opcodes were not present in the actual code.

File: test/libethereum/test/performance/rc5.sol [hunk 9]
```diff
@@ -0,0 +1,101 @@
+// https://people.csail.mit.edu/rivest/Rivest-rc5rev.pdf
```
Comment (chriseth) [inline test/libethereum/test/performance/rc5.sol]: Hm, wasn't RC5 patented?

File: libevm/VM.cpp (hunk 1 above)
Comment (gcolvin) [inline libevm/VM.cpp]: Yes.  They were locals, and will be again if possible.

File: libevm/VM.cpp (hunk 2 above)
Comment (gcolvin) [inline libevm/VM.cpp]: fixed

File: libevm/VM.cpp (hunk 3 above)
Comment (gcolvin) [inline libevm/VM.cpp]: fixed

File: libevm/VMOpt.cpp (hunk 4 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: If you mean 112-114, yes, it's tricky code, and I suspect it's wrong, thanks. (Am debugging behind the config macros).  There is similar code in VM.cpp but it operates directly on the program counter and stack pointer.

File: libevmcore/Instruction.cpp (hunk 6 above)
Comment (gcolvin) [inline libevmcore/Instruction.cpp]: I thought so too, but they need to be used like regular opcodes, so they need to be in the same tables.  If ever new opcodes conflict there will be plenty of time to reassign these internally.  We could make an EIP issue to reserve a range of opcodes for this purpose, but that could delay this thing by months.  Currently these are behind a config macro that is off.

File: libevm/VMOpt.cpp (hunk 5 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: May be I'm missing something, but the outer loop is over the whole code vector - `O(n)`  The inner loop for the constant pool is over at most 32 bytes.  The `verifyJumpDest()` lookup for the jump replacement is log the number of jump destinations.  So `not n log n` unless the code is nothing but jump destinations.

File: libevmcore/Instruction.h (hunk 7 above)
Comment (gcolvin) [inline libevmcore/Instruction.h]: I can't find a problem in my editor, even with tabs made visible and counting them.  I does look bad on github though.

File: libevm/VMOpt.cpp (hunk 8 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Hmm.  We can't throw the invalid opcode exception until it is encountered at runtime.  I think I could check here that m_code[i] is legit, and if not just leave it there.  Will make that change.

File: test/libethereum/test/performance/rc5.sol (hunk 9 above)
Comment (gcolvin) [inline test/libethereum/test/performance/rc5.sol]: How about that.  Fair use?  It's only performance test code, not even debugged yet.

**Comment by gcolvin** [comment]
Reviewers should double-check, but I think every non-style problem is inside configuration macros that are off.  @bobsummerwill suggested using macros so I could merge in-progress code without waiting so long between PR's.

**Comment by gcolvin** [comment]
Wrong - the synthetic opcodes are not just style. i do think they are OK.

File: libevm/VM.cpp (hunk 1 above)
Comment (chriseth) [inline libevm/VM.cpp]: Please use `m_pc`, members are members.

File: libevm/VMOpt.cpp (hunk 5 above)
Comment (chriseth) [inline libevm/VMOpt.cpp]: We have to protect against the worst case which is, as you say, "the code is nothing but jump destinations". You can leave that in if it is disabled by default, but then please put a comment.

File: libevmcore/Instruction.cpp (hunk 6 above)
Comment (chriseth) [inline libevmcore/Instruction.cpp]: What about prefixing them? Oh, now I see that it actually says `Instruction::JUMP`, is that intended?

File: libevmcore/Instruction.h (hunk 7 above)
Comment (chriseth) [inline libevmcore/Instruction.h]: `git checkout -p origin/develop Instruction.h` provides you with an interactive tool that lets you revert single chunks.

File: libevm/VMOpt.cpp (hunk 8 above)
Comment (chriseth) [inline libevm/VMOpt.cpp]: Ok! Then again: please put a comment.

File: test/libethereum/test/performance/rc5.sol (hunk 9 above)
Comment (chriseth) [inline test/libethereum/test/performance/rc5.sol]: I'm european, for me software patents do not exist ;-)

File: libevm/VM.cpp (hunk 3 above)
Comment (chriseth) [inline libevm/VM.cpp]: I don't see the fix here.

File: libevm/VM.cpp [hunk 10]
```diff
@@ -151,764 +136,745 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
...
 			updateIOGas();
 
-			*++m_SP = m_ext->envInfo().gasLimit();
-			break;
+			--SP;
+			++PC;
+		CASE_END
 
-		case Instruction::POP:
+		CASE_BEGIN(PUSHC)
```
Comment (chriseth) [inline libevm/VM.cpp]: Shouldn't that be "macroed out"?

File: libevm/VM.cpp [hunk 11]
```diff
@@ -151,764 +136,745 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
...
 			onOperation();
 			updateIOGas();
 
-			--m_SP;
-			break;
+			*++SP = m_pool[m_code[++PC]];
+			++PC;
```
Comment (chriseth) [inline libevm/VM.cpp]: Are you sure about this increment?

File: libevm/VM.cpp [hunk 12]
```diff
@@ -151,764 +136,745 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
...
-		case Instruction::DUP15:
-		case Instruction::DUP16:
+			if (*(SP - 1))
+				PC = verifyJumpDest(*SP);
+			else
+				++PC;
+			SP -= 2;
+		CASE_END
+
+		CASE_BEGIN(JUMPV)
```
Comment (chriseth) [inline libevm/VM.cpp]: Again: macroed out? (also for JUMPVI)

File: libevm/VM.cpp [hunk 13]
```diff
@@ -151,764 +136,745 @@ bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
...
 			onOperation();
 			updateIOGas();
 
-			auto n = (unsigned)m_inst - (unsigned)Instruction::SWAP1 + 2;
-			auto d = *m_SP;
-			*m_SP = m_stack[(1 + m_SP - m_stack) - n];
-			m_stack[(1 + m_SP - m_stack) - n] = d;
-			break;
+			auto n = (unsigned)INST - (unsigned)Instruction::SWAP1 + 2;
+			auto d = *SP;
```
Comment (chriseth) [inline libevm/VM.cpp]: Why not use `std::swap` here?

File: libevm/VMCalls.cpp [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+using namespace dev;
+using namespace dev::eth;
+
+
+
+void VM::copyDataToMemory(bytesConstRef _data, u256*& SP)
+{
+	auto offset = static_cast<size_t>(*SP--);
+	s512 bigIndex = *SP--;
+	auto index = static_cast<size_t>(bigIndex);
```
Comment (chriseth) [inline libevm/VMCalls.cpp]: No need for bigindex here, right?

File: libevm/VMConfig.h [hunk 15]
```diff
@@ -0,0 +1,294 @@
+///////////////////////////////////////////////////////////////////////////////
```
Comment (chriseth) [inline libevm/VMConfig.h]: License header

File: libevm/VMOpt.cpp [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+	m_bounce = &VM::interpretCases;	
+	m_code_vector = m_ext->code;
+	m_code = m_code_vector.data();
+	
+	initMetrics();	
+	
+	optimize();
+}
+
+int VM::pool_constant(const u256& _con)
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: `u256 const&`

File: libevm/VMOpt.cpp [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+{
+	int i = 0, n = m_pool_vector.size();
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: Name should be `m_poolVector`, also the type should not be part of the name. Perhaps it makes sense to single out the constant pool into its own class?

File: libevm/VMOpt.cpp [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+	while (i < n)
```
Comment (chriseth) [inline libevm/VMOpt.cpp]: I know this is not finished code, but this needs about 60000 comparisons of `u256` variables in the worst case. Does that take long? You probably convert the input variable to a `uint8_t` early on, so probably not an issue.

File: libevmcore/Instruction.cpp (hunk 6 above)
Comment (gcolvin) [inline libevmcore/Instruction.cpp]: Yes, they have to look like instructions to the interpreter, not just numbers.  The only place they are used outside of the interpreter is tracing and debugging via onOperation().   There is a comment in Instruction.h, and I just put them inside macros:
```
#ifdef EVM_USE_SYNTHETIC_INSTRUCTIONS
	// these are generated by the interpreter - should never be in user code
	PUSHC = 0xad,		///< push value from constant pool
	JUMPV,				///< alter the program counter - pre-verified
	JUMPVI,				///< conditionally alter the program counter - pre-verified
#endif

```

File: libevm/VMOpt.cpp (hunk 8 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: This may turn out to be the most heavily comment module in the system ;)

File: libevm/VM.cpp (hunk 10 above)
Comment (gcolvin) [inline libevm/VM.cpp]: Yes, doing that.

File: libevm/VMOpt.cpp (hunk 5 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Comment in.  Jeffrey found it to be an optimization worth doing in the go interpreter.
  https://blog.ethereum.org/2016/06/02/go-ethereums-jit-evm/

File: libevmcore/Instruction.h (hunk 7 above)
Comment (gcolvin) [inline libevmcore/Instruction.h]: Problem was wrong tab stops in my editor.  There was some other misaligned stuff already there, so I tried to fix that to.

File: libevm/VM.cpp (hunk 13 above)
Comment (gcolvin) [inline libevm/VM.cpp]: Existing, working code I didn't consider changing.

File: libevm/VMCalls.cpp (hunk 14 above)
Comment (gcolvin) [inline libevm/VMCalls.cpp]: Not any more.  I still need to review the code for every remaining place with a too-wide type, and for uint64_t values passed to functions that take a size_t.  (You changed some size_t a while back.)

File: libevm/VMConfig.h (hunk 15 above)
Comment (gcolvin) [inline libevm/VMConfig.h]: Aha.

File: libevm/VMOpt.cpp (hunk 16 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Thanks.  30 year old habits extinguish slowly.

File: libevm/VMOpt.cpp (hunk 17 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Fixed the name and some more like it.  There are three vectors using this pattern, not sure that they are worth a class.
	// space for code and pointer to data
	bytes mCodeSpace;
	byte* mCode;

	// space for stack and pointer to data
	u256s mStackSpace;
	u256* mStack;
	
	// space for constant pool and pointer to data
	u256s mPoolSpace;
	u256* mPool;
```

File: libevm/VMOpt.cpp (hunk 18 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: if you mean `if (mPoolSpace[i] == _con)` then both are u256.   But I can reduce the number of comparisons by using a map to find the unique constants, then copy it to a vector when done.  Another possibility is to wait until encountering a PUSHn to put the constant in the pool and replace the PUSHn with PUSHC.  Once it's working I can look at that.

File: libevm/VM.cpp (hunk 10 above)
Comment (gcolvin) [inline libevm/VM.cpp]: Done.

File: libevm/VM.cpp (hunk 11 above)
Comment (gcolvin) [inline libevm/VM.cpp]: I think it's wrong, thank s.  Still debugging this code, and it should be inside macros.

**Comment by gcolvin** [comment]
And thanks @chriseth and @winsvega for the thorough review!

File: libevm/VM.cpp (hunk 3 above)
Comment (gcolvin) [inline libevm/VM.cpp]: hadn't pushed yet

File: libevm/VMOpt.cpp (hunk 17 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Went ahead and fixed them, as Travis is just sitting there anyway.

File: libevm/VM.cpp (hunk 1 above)
Comment (gcolvin) [inline libevm/VM.cpp]: Just noticed this, sorry.  Will be a separate commit.

File: libevm/VMOpt.cpp (hunk 17 above)
Comment (gcolvin) [inline libevm/VMOpt.cpp]: Fixed them.  Travis hadn't even started.

File: libevm/VMOpt.cpp (hunk 5 above)
Comment (chriseth) [inline libevm/VMOpt.cpp]: I am not 100% sure but in that article he talks about doing jit optimizations in the background, so in their case, it does not hurt too much if these optimizations take way too long, because the code is already executed in the foreground.

Comment (gcolvin) [inline libevm/VMOpt.cpp]: I mean that it performs well enough to be worth doing.

File: libevmcore/Instruction.cpp [hunk 19]
```diff
@@ -329,13 +334,13 @@ string dev::eth::disassemble(bytes const& _mem)
...
 		if (!isValidInstruction(_instr))
-			ret << "0x" << hex << int(_instr) << " ";
+		ret << "0x" << hex << int(_instr) << " ";
 		else
 		{
-			InstructionInfo info = instructionInfo(_instr);
-			ret << info.name << " ";
-			if (info.additional)
-				ret << "0x" << hex << _data << " ";
+		InstructionInfo info = instructionInfo(_instr);
```
Comment (chriseth) [inline libevmcore/Instruction.cpp]: Indentation looks wrong

File: libevmcore/Instruction.cpp [hunk 20]
```diff
@@ -312,13 +317,13 @@ void dev::eth::eachInstruction(
...
 		if (isValidInstruction(instr))
-			additional = instructionInfo(instr).additional;
+		additional = instructionInfo(instr).additional;
 		u256 data;
 		for (size_t i = 0; i < additional; ++i)
 		{
-			data <<= 8;
-			if (++it < _mem.end())
-				data |= *it;
+		data <<= 8;
```
Comment (chriseth) [inline libevmcore/Instruction.cpp]: Indentation looks wrong

File: libevm/VM.cpp [hunk 21]
```diff
@@ -71,17 +47,24 @@ template <class S> S modWorkaround(S const& _a, S const& _b)
...
+		if (*m_onOp)
+			(*m_onOp)(++m_nSteps, m_pc, m_op,
+				m_newMemSize > m_mem.size() ? (m_newMemSize - m_mem.size()) / 32 : uint64_t(0),
+				m_runGas, *m_io_gas, this, m_ext);
+	}
+#else
+	void VM::onOperation()
+	{
+	}
+	#define onOperation() ()
```
Comment (chriseth) [inline libevm/VM.cpp]: I think this causes a build error

**Comment by gcolvin** [comment]
@chriseth @winsvega @bobsummerwill  Can we get this merged?  I need it in before the next stuff can go in.  It builds and tests clean in the default configuration, and builds in all configurations.
Christian:  I missed the indentation problems you noticed yesterday - looks like my editor smashed a few tabs out.  I've fixed it in my branch for next time.

**Comment by bobsummerwill** [comment]
If we're down to whitespace, then please let's merge this tomorrow, eh?

**Comment by chriseth** [comment]
@gcolvin sure, indentation and build error were the only blockers.
